integrations/kube-agent-updater/pkg/controller/statefulset.go

```diff
@@ -0,0 +1,284 @@
+	}
+
+	// Get the current and past version
+	currentVersion, err := getWorkloadVersion(obj.Spec.Template.Spec)
+	if err != nil {
+		switch trace.Unwrap(err).(type) {
+		case *trace.BadParameterError:
+			log.Info("Teleport container found, but failed to get version from the img tag. Will continue and do a version update.")
+		default:
+			return requeueLater, trace.Wrap(err)
+		}
+	}
+
+	// Check if the statefulset is blocked by an old unhealthy pod, if so
+	// attempt to remove it. To ensure idempotentcy this happens also outside
+	// of maintenance windows and if no update is available.
+	// Deferring the unblock allows it to happen after a potential update. We
+	// won't need to wait another reconciliation window, or 10+ minutes for new
+	// pods to be created from the new StatefulSet spec.
+	defer func() {
+		// Note: this might be called often, we might decide to enable cache for pods
+		// for performance purposes (we trade more memory usage on the updater side,
+		// against less latency and calls to the apiserver)
+		if err := r.unblockStatefulSetRollout(ctx, &obj); err != nil {
+			log.Error(err, "statefulset unblocking failed, the rollout might get stuck")
+		}
+	}()
+
+	image, err := r.GetVersion(ctx, &obj, currentVersion)
+	var (
+		noNewVersionErr *NoNewVersionError
+		maintenanceErr  *MaintenanceNotTriggeredError
+	)
+	switch {
+	case errors.As(err, &noNewVersionErr):
+		// Error contains the detected versions
+		log.Info("StatefulSet is already up-to-date, not updating.", "err", err)
+		return requeueLater, nil
+	case errors.As(err, &maintenanceErr):
+		// Not logging the error because it provides no other information than its type.
+		log.Info("No maintenance triggered, not updating.", "currentVersion", currentVersion)
+		return requeueLater, nil
+	case trace.IsTrustError(err):
+		// Logging as error as image verification should not fail under normal use
+		log.Error(err, "Image verification failed, not updating.")
+		return requeueLater, trace.Wrap(err)
+	case err != nil:
+		log.Error(err, "Unexpected error, not updating.")
+		return requeueLater, trace.Wrap(err)
+	}
+
+	log.Info("Updating podSpec with image", "image", image.String())
+	err = setContainerImageFromPodSpec(&obj.Spec.Template.Spec, teleportContainerName, image.String())
+	if err != nil {
+		return requeueLater, trace.Wrap(err)
+	}
+
+	if err = r.Update(ctx, &obj); err != nil {
+		return requeueNow, trace.Wrap(err)
+	}
+	return requeueLater, nil
+}
+
+// SetupWithManager makes the DeploymentVersionUpdater managed by a ctrl.Manager.
+// Once started, the manager will send Deployment reconciliation requests to the
+// DeploymentVersionUpdater controller.
+func (r *StatefulSetVersionUpdater) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appsv1.StatefulSet{}).
+		Complete(r)
+}
+
+func (r *StatefulSetVersionUpdater) unblockStatefulSetRollout(ctx context.Context, sts *appsv1.StatefulSet) error {
```

Comment:
Nit: `maybeUnblockStatefulSetRollout` would probably be a bit more appropriate name for this (and adding a godoc explaining what it does wouldn't hurt either).